tools/autograd/gen_inplace_or_view_type.py

```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
                      '/* is_fw_differentiable */ true, '
                      f'/* creation_meta */ {creation_meta});')
             rhs_value = f'std::move({var})'
         else:
             _, unpacked_bindings = unpack_args(f)
             call += emit_view_lambda(f, unpacked_bindings)
-            creation_meta = 'GradMode::is_enabled() ? CreationMeta::DEFAULT: CreationMeta::NO_GRAD_MODE'
-            rhs_value = (f'as_view(/* base */ {view_info}, /* output */ {var}, /* is_bw_differentiable */ true, '
+            creation_meta = ('at::GradMode::is_enabled() ? '
+                             'torch::autograd::CreationMeta::DEFAULT : '
+                             'torch::autograd::CreationMeta::NO_GRAD_MODE')
+            rhs_value = (f'torch::autograd::as_view(/* base */ {view_info}, /* output */ {var}, /* is_bw_differentiable */ true, '
                          '/* is_fw_differentiable */ true, '
                          f'/* view_func */ func, /* creation_meta */ {creation_meta})')
     else:
         # This could be supported but we don't need it at the moment, so keeping things simple.
         raise RuntimeError('Function that return multiple differentiable output '
                            'when at least one of them is view is not supported.')
     return call, rhs_value
+
+def emit_inplace_or_view_body(fn: NativeFunctionWithDifferentiabilityInfo) -> List[str]:
+    f = fn.func
+    inplace_view_body: List[str] = []
+
+    dispatcher_sig = DispatcherSignature.from_schema(f.func)
+    dispatcher_exprs = dispatcher_sig.exprs()
+
+    ret_and_arg_types = ', '.join([dispatcher_sig.returns_type()] + [a.type.cpp_type() for a in dispatcher_exprs])
+    # code-generated tracing kernels plumb and recompute dispatch keys directly through the kernel for performance.
+    # See Note [Plumbing Keys Through The Dispatcher] for details.
+    dispatch_key_set = 'ks & c10::DispatchKeySet(c10::DispatchKeySet::FULL_AFTER, c10::DispatchKey::InplaceOrView)'
+    redispatch_args = ', '.join([dispatch_key_set] + [a.expr for a in dispatcher_exprs])
+
+    # Note that this calls the slow, dispatching variants of manual_cpp_binding ops.
+    # We could probably work harder to ensure that the fast variants are called instead, but the perf benefit would be minimal.
+    sig_group = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=f.manual_cpp_binding)
+    if sig_group.faithful_signature is not None:
+        api_name = sig_group.faithful_signature.name()
+    else:
+        api_name = sig_group.signature.name()
+    # inplace_view_body.append(THROW_IF_VARIABLETYPE_ON)
+    if modifies_arguments(f):  # inplace op
+        inplace_view_body.append(INPLACE_REDISPATCH.substitute(
+            api_name=api_name,
+            unpacked_args=redispatch_args,
+        ))
+        for r in cpp.return_names(f):
+            inplace_view_body.append(f'torch::autograd::increment_version({r});')
```

Comment:
This is unsound for functions which mutate input arguments but don't return them. We even have a list of these:

```
        if self.name.name.inplace:
            # TODO: fixme
            if str(self.name) not in [
                    '_amp_foreach_non_finite_check_and_unscale_',
                    '_foreach_add_.ScalarList',
                    '_foreach_sub_.ScalarList',
                    '_foreach_mul_.ScalarList',
                    '_foreach_div_.ScalarList',
                    '_foreach_add_.Scalar',
                    '_foreach_sub_.Scalar',
                    '_foreach_mul_.Scalar',
                    '_foreach_div_.Scalar',
                    '_foreach_add_.List',
                    '_foreach_sub_.List',
                    '_foreach_mul_.List',
                    '_foreach_div_.List',
                    '_foreach_exp_',
                    '_foreach_sqrt_',
                    '_foreach_abs_',
                    '_foreach_acos_',
                    '_foreach_asin_',
                    '_foreach_atan_',
                    '_foreach_ceil_',
                    '_foreach_cos_',
                    '_foreach_cosh_',
                    '_foreach_erf_',
                    '_foreach_erfc_',
                    '_foreach_expm1_',
                    '_foreach_floor_',
                    '_foreach_log_',
                    '_foreach_log10_',
                    '_foreach_log1p_',
                    '_foreach_log2_',
                    '_foreach_neg_',
                    '_foreach_tan_',
                    '_foreach_tanh_',
                    '_foreach_sin_',
                    '_foreach_sinh_',
                    '_foreach_round_',
                    '_foreach_lgamma_',
                    '_foreach_frac_',
                    '_foreach_reciprocal_',
                    '_foreach_sigmoid_',
                    '_foreach_trunc_',
                    '_foreach_addcmul_.Scalar',
                    '_foreach_addcdiv_.Scalar',
                    '_foreach_addcmul_.ScalarList',
                    '_foreach_addcdiv_.ScalarList',
                    '_foreach_zero_']:
                assert len(self.returns) == 1
```

Reply:
But I guess this logic was taken from the preexisting autograd codegen, so it is technically no worse. Should file an issue for this.

Reply:
I think that none of these support autograd.
So VariableType should just be a no-op.

In this case, I don't think you need to generate a kernel for them because right now, you generate something pretty empty: https://ezyang.github.io/pytorch-ci-hud/build1/pytorch-master

Reply:
If we ever want to differentiate through optimizer updates, the foreach kernels ought to support autograd eventually.

Reply:
Good point! I excluded these ops in InplaceOrView kernels and created an issue in https://github.com/pytorch/pytorch/issues/53796.